host/switch.go

```diff
@@ -0,0 +1,189 @@
+	groupErrs, ctx := errgroup.WithContext(ctx)
+	defer os.Exit(exitCode)
+
+	// catch user issued signals
+	sigChan := make(chan os.Signal, 1)
+	signal.Notify(sigChan, os.Interrupt, syscall.SIGTERM, syscall.SIGINT)
+
+	subnet, err := validateSubnet(VirtualSubnet)
+	if err != nil {
+		logrus.Fatal(err)
+	}
+
+	portForwarding, err := parsePortForwarding(StaticPortForward)
+	if err != nil {
+		logrus.Fatal(err)
+	}
+
+	config := newConfig(*subnet, portForwarding, debug)
+
+	ln, err := vsockListener(vsockHandshakePort, vsockListenPort, seedPhrase)
+	if err != nil {
+		logrus.Fatalf("creating vsock listener for hostSwitch failed: %v", err)
+	}
+	logrus.Debugf("attempting to start a virutal network with the following config: %+v", config)
+	groupErrs.Go(func() error {
+		return run(ctx, groupErrs, &config, ln)
+	})
+
+	// Wait for something to happen
+	groupErrs.Go(func() error {
+		select {
+		// Catch signals so exits are graceful and defers can run
+		case <-sigChan:
+			cancel()
+			return errors.New("signal caught")
+		case <-ctx.Done():
+			return nil
+		}
+	})
+	// Wait for all of the go funcs to finish up
+	if err := groupErrs.Wait(); err != nil {
+		logrus.Error(err)
+		exitCode = 1
+	}
+}
+
+func run(ctx context.Context, g *errgroup.Group, config *types.Configuration, ln net.Listener) error {
+	vn, err := virtualnetwork.New(config)
+	if err != nil {
+		return err
+	}
+	logrus.Info("waiting for clients...")
+	go func() {
+		for {
+			conn, err := ln.Accept()
+			if err != nil {
+				logrus.Errorf("failed to accept: %v", err)
+			}
+			err = vn.AcceptQemu(ctx, conn)
+			if err != nil {
+				logrus.Error(errors.Wrap(err, "AcceptQemu failed"))
+			} else {
+				logrus.Infof("accepted connection: ctx=%+v conn=%+v", ctx, conn)
+			}
+		}
+	}()
+
+	apiServer := fmt.Sprintf("%s:80", config.GatewayIP)
+	vnLn, err := vn.Listen("tcp", apiServer)
+	if err != nil {
+		return err
+	}
+	mux := http.NewServeMux()
+	mux.Handle("/services/forwarder/all", vn.Mux())
+	mux.Handle("/services/forwarder/expose", vn.Mux())
+	mux.Handle("/services/forwarder/unexpose", vn.Mux())
+	httpServe(ctx, g, vnLn, mux)
+	logrus.Infof("port forwarding API server is running on: %s", apiServer)
+
+	if debug {
+		g.Go(func() error {
+		debugLog:
+			for {
+				select {
+				case <-time.After(5 * time.Second):
+					logrus.Debugf("%v sent to the VM, %v received from the VM\n", humanize.Bytes(vn.BytesSent()), humanize.Bytes(vn.BytesReceived()))
+				case <-ctx.Done():
+					break debugLog
+				}
+			}
+			return nil
+		})
+	}
+	return nil
+}
+
+func httpServe(ctx context.Context, g *errgroup.Group, ln net.Listener, mux http.Handler) {
+	g.Go(func() error {
+		<-ctx.Done()
+		return ln.Close()
+	})
+	g.Go(func() error {
+		s := &http.Server{
+			Handler:      mux,
+			ReadTimeout:  10 * time.Second,
+			WriteTimeout: 10 * time.Second,
+		}
+		err := s.Serve(ln)
+		if err != nil {
+			if err != http.ErrServerClosed {
+				return err
+			}
+			return err
```

Comment:
Wait, so we don't actually care if `err` is `http.ErrServerClosed` at all, we just always return `err`?